pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	dto "github.com/prometheus/client_model/go"
+
+	burrow "github.com/goburrow/cache"
+)
+
+// StatsCounter is an interface for recording cache stats for goburrow.Cache.
+// The StatsCounter can be found at
+// - https://github.com/goburrow/cache/blob/f6da914dd6e3546dffa8802919dbca80cd33abe3/stats.go#L67
+var _ burrow.StatsCounter = (*burrowStatsCounter)(nil)
+
+// burrowStatsCounter is a StatsCounter implementation for burrow cache.
+// It is a wrapper around prometheus metrics.
+// It has been intended to passed through the cache using cache.WithStatsCounter option
+// - https://github.com/goburrow/cache/blob/f6da914dd6e3546dffa8802919dbca80cd33abe3/local.go#L552
+type burrowStatsCounter struct {
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	hits     prometheus.Counter
+	miss     prometheus.Counter
+	eviction prometheus.Counter
+
+	trackLoadingCacheStats bool
+	loadSuccess            prometheus.Counter
+	loadError              prometheus.Counter
+	loadTotalTime          prometheus.Histogram
```

Comment:
do you think this is worth the trouble as it will always be in memory?

Reply:
It's not initialized when you don't use a loading cache (which we don't). But also, Prometheus histograms under the hood just counter traditionally. This shouldn't have significant memory pressure when it's enabled.

Reply:
it's ~10 counters, memory wise it's negligible